filament/src/MorphTargetBuffer.cpp

```diff
@@ -74,114 +74,110 @@ static inline size_t getHeight(size_t vertexCount) noexcept {
-            stride, getHeight(vertexCount), 1);
-}
-
-static inline size_t getPositionIndex(size_t targetIndex) noexcept {
-    return targetIndex * 2 + 0;
+            stride, height, 1);
 }
 
-static inline size_t getTangentIndex(size_t targetIndex) noexcept {
-    return targetIndex * 2 + 1;
+template<>
+inline size_t getSize<VertexAttribute::TANGENTS>(size_t vertexCount) noexcept {
+    const size_t stride = getWidth(vertexCount) * sizeof(short4);
+    const size_t height = getHeight(vertexCount);
+    return Texture::PixelBufferDescriptor::computeDataSize(
+            Texture::PixelBufferDescriptor::PixelDataFormat::RGBA_INTEGER,
+            Texture::PixelBufferDescriptor::PixelDataType::SHORT,
+            stride, height, 1);
 }
 
 FMorphTargetBuffer::FMorphTargetBuffer(FEngine& engine, const Builder& builder)
         : mSBuffer(PerRenderPrimitiveMorphingSib::SAMPLER_COUNT),
           mVertexCount(builder->mVertexCount),
           mCount(builder->mCount) {
     FEngine::DriverApi& driver = engine.getDriverApi();
-
     mSbHandle = driver.createSamplerGroup(PerRenderPrimitiveMorphingSib::SAMPLER_COUNT);
-    mTbHandle = driver.createTexture(SamplerType::SAMPLER_2D_ARRAY, 1, TextureFormat::RGBA32F, 1,
-            getWidth(mVertexCount), getHeight(mVertexCount), getDepth(mCount),
-            TextureUsage::DEFAULT);
+    mPbHandle = driver.createTexture(SamplerType::SAMPLER_2D_ARRAY, 1, TextureFormat::RGB32F, 1,
+            getWidth(mVertexCount), getHeight(mVertexCount), mCount, TextureUsage::DEFAULT);
+    mTbHandle = driver.createTexture(SamplerType::SAMPLER_2D_ARRAY, 1, TextureFormat::RGBA16I, 1,
+            getWidth(mVertexCount), getHeight(mVertexCount), mCount,TextureUsage::DEFAULT);
 
     SamplerParams samplerParams{};
     samplerParams.filterMin = SamplerMinFilter::NEAREST;
     samplerParams.filterMag = SamplerMagFilter::NEAREST;
-    mSBuffer.setSampler(PerRenderPrimitiveMorphingSib::TARGETS, mTbHandle, samplerParams);
+    mSBuffer.setSampler(PerRenderPrimitiveMorphingSib::POSITIONS, mPbHandle, samplerParams);
+    mSBuffer.setSampler(PerRenderPrimitiveMorphingSib::TANGENTS, mTbHandle, samplerParams);
 }
 
 void FMorphTargetBuffer::terminate(FEngine& engine) {
     FEngine::DriverApi& driver = engine.getDriverApi();
-
     driver.destroySamplerGroup(mSbHandle);
     driver.destroyTexture(mTbHandle);
+    driver.destroyTexture(mPbHandle);
 }
 
 void FMorphTargetBuffer::setPositionsAt(FEngine& engine, size_t targetIndex, math::float3 const* positions, size_t count) {
     ASSERT_PRECONDITION(targetIndex < mCount, "targetIndex must be < count");
 
-    auto size = getSize(mVertexCount);
+    auto size = getSize<VertexAttribute::POSITION>(mVertexCount);
 
-    ASSERT_PRECONDITION((int)sizeof(math::float3) * count <= size,
-            "MorphTargetBuffer (size=%lu) overflow (size=%lu)",
-            size, sizeof(math::float3) * count);
+    ASSERT_PRECONDITION(sizeof(float3) * count <= size,
```

Comment:
maybe put `sizeof(float3)*count` into a local

Reply:
Ok it's done.